File: canser.py
"""
Supported data types:

- Integer (64 bits)
- Float (64 bits)
- String (UTF-8)
- List
- Dictionary

Little endianness is assumed.
"""

import struct


def save(x, fp):
    if isinstance(x, float):
        fp.write(b"f")
        fp.write(struct.pack("<d", x))
    elif isinstance(x, int):
        fp.write(b"i")
        _save_int(x, fp)
    elif isinstance(x, str):
        b = x.encode("utf-8")
        fp.write(b"s")
        _save_int(len(b), fp)
        fp.write(b)
    elif isinstance(x, list):
        fp.write(b"l")
        _save_int(len(x), fp)
        for v in x:
            save(v, fp)
    elif isinstance(x, dict):
        fp.write(b"d")
        _save_int(len(x), fp)
        for k in sorted(x.keys()):
            save(k, fp)
            save(x[k], fp)
    else:
        raise ValueError(f"Unsupported argument {x} of type {type(x)} for `save`")


def load(fp):
    tag = fp.read(1)
    if tag == b"f":
        return struct.unpack("<d", fp.read(8))[0]
    elif tag == b"i":
        return _load_int(fp)
    elif tag == b"s":
        return fp.read(_load_int(fp)).decode("utf-8")
    elif tag == b"l":
        return [load(fp) for _ in range(_load_int(fp))]
    elif tag == b"d":
        d = dict()
        for _ in range(_load_int(fp)):
            k = load(fp)
            v = load(fp)
            d[k] = v
        return d
    else:
        raise ValueError(f"Invalid tag {tag} for {fp}")


def _save_int(x, fp):
    return fp.write(struct.pack("<q", x))


def _load_int(fp):
    return struct.unpack("<q", fp.read(8))[0]

File: CanSer.jl
"""
Supported data types:

- Integer (64 bits)
- Float (64 bits)
- String (UTF8)
- Vector
- Dictionary

Little endiannness is assumed.
"""
module CanSer

save(io::IO, x::Float64) = write(io, 'f', x)
save(io::IO, x::Int64) = write(io, 'i', x)
save(io::IO, x::String) = write(io, 's', sizeof(x), x)

function save(io::IO, x::Vector)
    write(io, 'l', length(x))
    for v in x
        save(io, v)
    end
end

function save(io::IO, x::Dict)
    write(io, 'd', length(x))
    for k in sort(collect(keys(x)))
        save(io, k)
        save(io, x[k])
    end
end


function load(io::IO)
    tag = read(io, Char)
    if tag == 'f'
        read(io, Float64)
    elseif tag == 'i'
        read(io, Int64)
    elseif tag == 's'
        String(read(io, read(io, Int64)))
    elseif tag == 'l'
        [load(io) for _ in 1:read(io, Int64)]
    elseif tag == 'd'
        Dict((k = load(io); v = load(io); k=>v) for _ in 1:read(io, Int64))
    else
        throw(ArgumentError("Invalid tag $tag for $io"))
    end
end

end
